homeassistant/components/ring/switch.py

```diff
@@ -0,0 +1,115 @@
+"""This component provides HA switch support for Ring Door Bell/Chimes."""
+from datetime import timedelta
+import logging
+
+from homeassistant.components.switch import SwitchDevice
+
+from . import DATA_RING
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = timedelta(seconds=10)
+
+SIREN_ICON = 'mdi:alarm-bell'
+LIGHT_ICON = 'mdi:track-light'
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up a sensor for a Ring device."""
+    ring = hass.data[DATA_RING]
+    logging.warning('setting up')
+    switches = []
+
+    for device in ring.stickup_cams:  # ring.stickup_cams is doing I/O
+        if device.has_capability('siren'):
+            switches.append(SirenSwitch(hass, device))
+        if device.has_capability('light'):
+            switches.append(LightSwitch(hass, device))
+
+    add_entities(switches, True)
+
+
+class BaseRingSwitch(SwitchDevice):
+    """Represents a switch for controlling an aspect of a ring device"""
+    def __init__(self, hass, device, device_type):
+        self._device = device
+        self._device_type = device_type
+        self._unique_id = '{}-{}'.format(self._device.id, self._device_type)
+
+    @property
+    def name(self):
+        """Name of the device."""
+        return '{} {}'.format(self._device.name, self._device_type)
+
+    @property
+    def unique_id(self):
+        """Return a unique ID."""
+        return self._unique_id
+
+    def update(self):
+        """Get the siren status from ring"""
+        self._device.update()
```

Comment:
This worries me a bit because for every device I will be making an update call (so if a device has a camera and a floodlight then there will be 2 calls) every 10 seconds. I'm guessing the sensors will be doing similar. I've taken a look at the code this calls (https://github.com/tchellomello/python-ring-doorbell/blob/master/ring_doorbell/generic.py#L52) and there isn't any sort of caching or anything to stop rapid calls here. I'm just not sure of a better way of doing this, or if it's just not a problem I should worry about?

Reply:
The pattern we normally use, is to move the data fetching to a single data instance that handles all the device or service interactions.

If there's no cache in the library we could build our own cache. Then give access to the data instance with cache to all entities and either let them pull state from the cache via the builtin home assistant polling mechanism or turn of polling and signal the entities to pull state after updating the cache via custom scheduling. If the former builtin polling mechanism is used, there needs to be a throttle on the data instance update method, so that just the first entity that calls the method will actually update the data and cache, within a time interval.